src/java/main/org/apache/zookeeper/ClientCnxn.java

```diff
@@ -1495,10 +1504,21 @@ public ReplyHeader submitRequest(RequestHeader h, Record request,
         Packet packet = queuePacket(h, r, request, response, null, null, null,
                 null, watchRegistration, watchDeregistration);
         synchronized (packet) {
+            long waitStartTime = System.currentTimeMillis();
             while (!packet.finished) {
-                packet.wait();
+                packet.wait(requestTimeout);
+                if (!packet.finished && ((System.currentTimeMillis()
```

Comment:
Yes, and actually we have a Time class which we can use here that uses nanoTime internally for exact same reason.